### PR TITLE
Adds MachineConfig Operator

### PR DESCRIPTION
**Pull request by abhinavdahiya**
* lib/resourceread,resourcemerge,resourceapply are sourced from https://github.com/openshift/library-go/tree/master/pkg/operator/resource and changed for local use.
* manifest/ has managed manifests. (these are stored as generated code in `pkg/operator/assets/bindata.go`)
* install/ has assets that install operator.  (these are stored as generated code in `pkg/operator/assets/bindata.go`)
* pkg/operator adds inits sync loop for MCO
* cmd/machine-config-operator adds the command machine-config-operator
* add 2-stage dockerfiles for machine-config-operator,controller

Follow up:
* Bootstrap phase for MCO
* tests for MCO
* Remove hard-coded controllerconfig when installconfig is available

### CODE REVIEW

File: pkg/operator/assets/bindata.go [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+	for _, child := range children {
+		err = RestoreAssets(dir, filepath.Join(name, child))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func _filePath(dir, name string) string {
```
Comment (yifan-gu) [inline pkg/operator/assets/bindata.go]: what is this for?

File: pkg/operator/sync.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	etcd := resourceread.ReadMachineConfigPoolV1OrDie(etcdBytes)
+	_, _, err = resourceapply.ApplyMachineConfigPool(optr.client.MachineconfigurationV1(), etcd)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (optr *Operator) syncMachineConfigController(config renderConfig) error {
+	ccBytes, err := renderAsset(config, "manifests/machineconfigcontroller/controllerconfig.yaml")
```
Comment (yifan-gu) [inline pkg/operator/sync.go]: same here, we can create a map from the asset name to their `render` and `read` functions, and potentially wait for the resources. Should be pretty generic to reduce the code redundancy .

File: pkg/operator/sync.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+		return err
+	}
+	if updated {
+		return optr.waitForDeploymentRollout(mcc)
+	}
+	return nil
+}
+
+func (optr *Operator) syncMachineConfigDaemon(config renderConfig) error {
+	mcdBytes, err := renderAsset(config, "manifests/machineconfigdaemon/daemonset.yaml")
```
Comment (yifan-gu) [inline pkg/operator/sync.go]: ditto here

**Comment by yifan-gu** [comment]
@abhinavdahiya LGTM overall, but seems a lot of the reconcile code can be compressed

File: pkg/operator/sync.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+	errs = append(errs, optr.syncMachineConfigDaemon(*config))
+
+	agg := utilerrors.NewAggregate(errs)
+	if agg != nil {
+		return fmt.Errorf("error syncing: %v", agg.Error())
+	}
+	return nil
+}
+
+func (optr *Operator) syncCustomResourceDefinitions(config renderConfig) error {
```
[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply pkg/operator/sync.go]: :+1: done

File: pkg/operator/sync.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+	if updated {
+		if err := optr.waitForCustomResourceDefinition(mcp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (optr *Operator) syncMachineConfigPools(config renderConfig) error {
+	masterBytes, err := renderAsset(config, "manifests/master.machineconfigpool.yaml")
```
[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply pkg/operator/sync.go]: :+1: done

File: pkg/operator/sync.go (hunk 2 above)
Comment (abhinavdahiya) [reply pkg/operator/sync.go]: read* and apply* expect specific types. 
Will end up doing bunch of type assertions.

**Comment by abhinavdahiya** [comment]
@yifan-gu PTAL

**Pull request merged**